library/src/main/java/com/datatorrent/lib/state/managed/AbstractManagedStateImpl.java

```diff
@@ -0,0 +1,499 @@
+ *     files to bucket files.
+ *   </li>
+ *   <li>
+ *     {@link #bucketsFileSystem}: a bucket on disk is sub-divided into time-buckets. This manages meta-bucket
+ *     information (list of {@link BucketsFileSystem.TimeBucketMeta}) per bucket.
+ *   </li>
+ *   <li>
+ *     {@link #timeBucketAssigner}: assigns time-buckets to keys and manages the time boundaries.
+ *   </li>
+ *   <li>
+ *     {@link #stateTracker}: tracks the size of data in memory and requests buckets to free memory when enough memory
+ *     is not available.
+ *   </li>
+ *   <li>
+ *     {@link #fileAccess}: plug-able file system abstraction.
+ *   </li>
+ * </ol>
+ *
+ * The implementations of put, getSync and getAsync here use windowId as the time field to derive timeBucket of a key.
+ */
+public abstract class AbstractManagedStateImpl
+    implements ManagedState, Component<OperatorContext>, Operator.CheckpointNotificationListener, ManagedStateContext,
+    TimeBucketAssigner.PurgeListener
+{
+  private long maxMemorySize;
+
+  protected int numBuckets;
+
+  @NotNull
+  private FileAccess fileAccess = new TFileImpl.DTFileImpl();
+  @NotNull
+  protected TimeBucketAssigner timeBucketAssigner = new TimeBucketAssigner();
+
+  protected Bucket[] buckets;
+
+  @Min(1)
+  private int numReaders = 1;
+  @NotNull
+  protected transient ExecutorService readerService;
+
+  @NotNull
+  protected IncrementalCheckpointManager checkpointManager = new IncrementalCheckpointManager();
+
+  @NotNull
+  protected BucketsFileSystem bucketsFileSystem = new BucketsFileSystem();
+
+  protected transient OperatorContext operatorContext;
+  protected transient long windowId;
+
+  @NotNull
+  protected Comparator<Slice> keyComparator = new SliceComparator();
+
+  protected final transient AtomicReference<Throwable> throwable = new AtomicReference<>();
+
+  @NotNull
+  @FieldSerializer.Bind(JavaSerializer.class)
+  private Duration checkStateSizeInterval = Duration.millis(
+      DAGContext.STREAMING_WINDOW_SIZE_MILLIS.defaultValue * OperatorContext.APPLICATION_WINDOW_COUNT.defaultValue);
+
+  @FieldSerializer.Bind(JavaSerializer.class)
+  private Duration durationPreventingFreeingSpace;
+
+  private transient StateTracker stateTracker = new StateTracker();
+
+  //accessible to StateTracker
+  final transient Object commitLock = new Object();
+
+  @Override
+  public void setup(OperatorContext context)
+  {
+    operatorContext = context;
+    fileAccess.init();
+
+    timeBucketAssigner.setPurgeListener(this);
+
+    //setup all the managed state components
+    timeBucketAssigner.setup(this);
+    checkpointManager.setup(this);
+    bucketsFileSystem.setup(this);
+
+    if (buckets == null) {
+      //create buckets array only once at start when it is not created.
+      numBuckets = getNumBuckets();
+      buckets = new Bucket[numBuckets];
+    }
+    for (Bucket bucket : buckets) {
+      if (bucket != null) {
```

Comment:
Yes it is possible. Let's say user specified numBuckets = 4. This will create ```buckets``` array of size 4. However if we don't see a key for bucket 3 then ```buckets[2]``` will never be initialized to a a bucket. We are following a lazy initialization strategy.